src/lib/warn-upload-overwrite.js

```diff
@@ -0,0 +1,146 @@
+  let localDoc = JSON.parse(JSON.stringify(doc));
+
+  // Pull remote _rev
+  let remoteRev;
+  let remoteDoc;
+  try {
+    remoteDoc = await db.get(localDoc._id);
+    remoteRev = remoteDoc._rev;
+  } catch (e) {
+    // continue regardless of error
+    log.trace('Trying to fetch remote _rev', e);
+  }
+
+  // Pull local _rev
+  let localRev;
+  try {
+    localRev = JSON.parse(fs.read(`${projectDir}/._revs/${localDoc._id}.json`).trim())[getRevsDocKey()];
+  } catch (e) {
+    // continue regardless of error
+    log.trace('Trying to fetch local _rev', e);
+  }
+
+  if (localRev) {
+    localDoc._rev = localRev;
+  }
+
+  if (localDoc._attachments) {
+    for (let key in localDoc._attachments) {
+      localDoc._attachments[key] = {
+        'content_type': doc._attachments[key].content_type, 
+        'revpos': localRev ? parseInt(localRev.split('-')[0]) : 0, 
+        'length': doc._attachments[key].data.length,
+        'digest': `md5-${crypto.createHash('md5').update(doc._attachments[key].data, 'binary').digest('base64')}`,
+        'stub': true
+      };
+    }
+  }
+
+  let diff; 
+  if (localDoc._id === 'settings' && remoteDoc._id === 'settings') {
+    diff = jsonDiff.diffString(remoteDoc.settings, localDoc.settings);
+  } else {
+    diff = jsonDiff.diffString(remoteDoc, localDoc);
+  }
+  if (diff) {  
+    if (!localRev) {
+      if (!readline.keyInYN(`We can't determine if you're going to overwrite someone else's changes or not. Would you like to proceed?`)) {
+        throw new Error('configuration modified');
+      }
+    } else if (localRev !== remoteRev) {
+      let index = readline.keyInSelect(responseChoicesWithDiff, question, {cancel: false});
+      if (index === 2) { // diff
+        log.info(diff);
+
+        index = readline.keyInSelect(responseChoicesWithoutDiff, question, {cancel: false});
+      }
+      if (index === 1) { // abort
+        throw new Error('configuration modified');
+      }
+    }
+  }
+
+  return doc;
+};
+
+const postUploadByRev = async (projectDir, db, doc) => {
+  const remoteDoc = await db.get(doc._id);
+
+  const revsDir = `${projectDir}/._revs`;
+  if (!fs.exists(revsDir)){
+    fs.mkdir(revsDir);
+  }
+
+  let revs = {};
+
+  const revsFile = `${revsDir}/${doc._id}.json`;
+  if (fs.exists(revsFile)) {
+    Object.assign(revs, JSON.parse(fs.read(revsFile).trim()));
+  }
+  revs[getRevsDocKey()] = remoteDoc._rev;
+
+  fs.write(revsFile, JSON.stringify(revs));
+
+  return doc;
+};
+
+const preUploadByXml = async (db, docId, localXml) => {
+  let remoteXml;
+  try {
+    const buffer = await db.getAttachment(docId, 'xml');
+    remoteXml = buffer.toString('utf8');
+  } catch (e) {
+    // continue regardless of error
+    log.trace('Trying to fetch remote xml', e);
+    throw e;
+  }
+
+  const localDom = new DOMParser().parseFromString(localXml);
+  const remoteDom = new DOMParser().parseFromString(remoteXml);
+
+  const diff = compare(localDom, remoteDom);
```

Comment:
Couch attachments have a hash we could store to compare later rather than comparing the raw xml. The xml is nice because unimportant changes (eg: whitespace) won't register as a diff, but the xml is generated so it should generate consistent output and therefore the hash should only fail for real errors.

Reply:
I think the issue here is that I need to compute the hash of the local xml and compare it with the hash created by couch. I have failed to replicate the algo used by couch. It seems that it is not just a straightforward md5sum followed by base64 encode. Couch db uses some compression of the xml as well: https://github.com/pouchdb/pouchdb/issues/3156#issuecomment-66831010

Reply:
Oh man that's a shame. Well it's ok to generate our own hash but I do think we need to do it in the post upload so we can only warn if the xml has been changed by someone else.